[Attention] Implement a proper fusion between first gemm and softmax

manupak opened the pull request:
The current code for generating attention kernels only accepted two optional inputs
that could be used to mul (scale) and add (bias) between the first gemm and the softmax.

However, going forward we need squeeze in a arbitrary dequantization schemes that is
composed of element wise operations. Therefore, as a pre-requisite to that this PR
implements a generic fusion point between the first gemm and softmax.

The said tree is implemented as an inverted tree where the root of the tree is the softmax input
and one leaf being the first gemm output. Other leaves can reach farther as long as elementwise operations
go. Moreover, this PR add support to fused reshape-like ops as well because they are quite common
in graphs we get from MIGraphX that has rank > 3 and we do squeeze them to be rank = 3 around gemms because
tosa only support rank=3 matmuls. 

Future work : 

However, the code done here fundamentally can fuse any operator that lowers
to rock.transform, into the tree with 1 restriction : the path between the first gemm output leaf to the next branching point
in the tree should contain an invertible rock.transform.

closes : https://github.com/ROCm/rocMLIR-internal/issues/1368

manupak committed:
* define attentionV2 with a region that describes a fusion point
* handle the tosa-to-rock lowering of that

manupak committed:
* folded the changes back to attention op (removed v2)
* end to end kernel pipeline is producing something
* need to handle the region in gridwise lowering

manupak committed:
* changed gridwise gemm to blockwise gemm to align the linalg
  block in preSoftmax fusion point.
* TODO : add support for unary elementwise in tosa to rock.

manupak committed:
* add view like ops into the mix
* cleaned up tosa to rock
* added handling of unary ops
* TODO: untransfrom + invert la generic inputs to tileviews

manupak committed:
* added handling of reshapes inside elemwise tree
* added more tests to cover elemwise fusion cases

manupak committed:
Merge branch 'develop' of https://github.com/ROCmSoftwarePlatform/rocMLIR into attn-presoftmax-fusion-point

manupak committed:
* remove some deadcode

manupak committed:
* clang format

manupak committed:
* remove commented code

manupak commented on mlir/include/mlir/Dialect/Rock/IR/RockOps.td:
```diff
@@ -444,8 +442,9 @@ def Rock_GridwiseAttentionAccelOp :
+  let regions = (region AnyRegion:$preSoftmaxBody);
   let assemblyFormat = [{
-    `(` operands `)` `features` `=` $features attr-dict `:` type(operands)
+    `(` operands `)` `features` `=` $features `preSoftmaxOps` `=` $preSoftmaxBody attr-dict `:` type(operands)
```
Ive intentionally left this like this to get ideas for a better assembly format.

krzysz00 replied:
Hm ... what's the usual mathematical syntax for attention?

The first observation I have is that you can use optional groups, the syntax being something like
```
(`preSoftmaxArgs` $preSoftmaxArgs^)?
```
or even
```
(`(` $preSoftmaxArgs^`)`)?
```

sjw36 replied:
I like the linalg.generic like approach. Can you follow that format more closely

manupak replied:
ok, I came up with the following : 

```mlir
    rock.attention{
     qk = %arg0 * %arg1 : memref<1x1024x32xf32>, memref<1x32x1024xf32>
     qk = elementwise otherIns(%arg3 : memref<1x1024x1024xf32>) {
    ^bb0(%arg5: memref<1x1024x1024xf32>, %arg6: memref<1x1024x1024xf32>, %arg7: memref<1x1024x1024xf32>):
      %0 = bufferization.to_tensor %arg5 restrict : memref<1x1024x1024xf32>
      %1 = bufferization.to_tensor %arg6 restrict : memref<1x1024x1024xf32>
      %2 = "tosa.mul"(%0, %1) <{shift = 0 : i32}> : (tensor<1x1024x1024xf32>, tensor<1x1024x1024xf32>) -> tensor<1x1024x1024xf32>
      %3 = bufferization.to_memref %2 : memref<1x1024x1024xf32>
      memref.copy %3, %arg7 : memref<1x1024x1024xf32> to memref<1x1024x1024xf32>
      rock.yield
    }
     %arg4 = softmax(qk) * %arg2 : memref<1x1024x32xf32> -> memref<1x1024x32xf32>
    } {arch = "amdgcn-amd-amdhsa:gfx90a:sramecc+:xnack-", features = #rock<GemmFeatures mfma|dot|atomic_add>}
```

I think it gives a clearer picture of the semantics of the op.

manupak committed:
* add gridwise attention op verifier to check #linalg ops

krzysz00 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -715,6 +717,47 @@ struct CollapseExpandRewritePattern
+        tosa::GreaterOp,
+        tosa::GreaterEqualOp
+       >(op);
+  // clang-format on
```
... I'm suspicious of this, but maybe this is the best we can do? Are we sure there isn't an interface somewhere?

manupak replied:
https://github.com/ROCm/rocMLIR/blob/4251df1dda5522598e5556179dd4359bd8f007c8/external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp#L58-L106

I just borrowed it from here... I suspect @sjw36 might know if there is any better way to do this -- if any?

manupak committed:
* clang format

sjw36 reviewed (approved):
Nice generalization!

krzysz00 reviewed (commented):
... Huh, my comments didn't submit

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/BufferizableOpInterfaceImpl.cpp:
```diff
@@ -90,9 +91,16 @@ struct GemmLikeInterface
+    if (op->getNumRegions() == 1) {
+      auto newOp = cloneWithoutRegions(
+          rewriter, op, /*newResultTypes=*/TypeRange{}, bufferArgs);
+      rewriter.inlineRegionBefore(op->getRegion(0), newOp->getRegion(0),
```
Is this right? Don't we want to clone the region?

manupak replied:
I was just basing this off from : https://github.com/ROCm/rocMLIR/blob/4251df1dda5522598e5556179dd4359bd8f007c8/external/llvm-project/mlir/lib/Dialect/Linalg/Transforms/BufferizableOpInterfaceImpl.cpp#L78-L81

but you maybe right, I ll try if we can remove inlineRegionBefore...

krzysz00 replied:
Huh, then that might just be it.

manupak replied:
it turned out i could just "clone" it.

krzysz00 commented on mlir/tools/rocmlir-gen/rocmlir-gen.cpp:
```diff
@@ -3584,17 +3625,18 @@ int main(int argc, char **argv) {
+    bufferizeOptions.disableRock = true;
+    rock::buildBufferizePipeline(pm, bufferizeOptions);
+
+    if (failed(pm.run(module))) {
```
Why're we making this unconditional here?
I'm a little bit dubious about putting this in `rocmlir-gen`. I'd rather rip this bit of code out and stick it as `-host-pipeline bufferization` over in `rocmlir-driver`

manupak replied:
Previously, we were only using tosa in host harness.. now we are using in rock ops themselves.
I dont see good reason why it needs to be conditional given that when we generating kernel or host code -- we rather prefer them to be buffered -- either by construction and/or lowering.

Im fine either way -- just wanted to keep the contract rocmlir-gen to always produce bufferized IR as it is currently.

krzysz00 replied:
Fair enough, reasonable point, let's not go aggressively refactoring things in this PR.

I'm thinking a future PR could be making `rocmlir-gen` output tensor code.

manupak committed:
* improve assembly format

manupak committed:
* cleaned up bufferization code

manupak commented:
I believe i have addressed the comments and Im taking this in to enable reviewing of https://github.com/ROCm/rocMLIR/pull/1411.

 So if you have more comments on this one, I m happy to address them in https://github.com/ROCm/rocMLIR/pull/1411 -- just to avoid the merging/rebasing pain from my side.
